src/execution/operator/join/physical_asof_join.cpp

```diff
@@ -0,0 +1,698 @@
+	}
+}
+
+unique_ptr<OperatorState> PhysicalAsOfJoin::GetOperatorState(ExecutionContext &context) const {
+	auto &config = ClientConfig::GetConfig(context.client);
+	return make_uniq<AsOfLocalState>(context.client, *this, config.force_external);
+}
+
+void PhysicalAsOfJoin::ResolveSimpleJoin(ExecutionContext &context, DataChunk &input, DataChunk &chunk,
+                                         OperatorState &lstate_p) const {
+	auto &lstate = lstate_p.Cast<AsOfLocalState>();
+	auto &gsink = sink_state->Cast<AsOfGlobalSinkState>();
+
+	// perform the actual join
+	bool found_match[STANDARD_VECTOR_SIZE] = {false};
+	lstate.ResolveJoin(input, found_match);
+
+	// now construct the result based on the join result
+	switch (join_type) {
+	case JoinType::MARK: {
+		PhysicalJoin::ConstructMarkJoinResult(lstate.lhs_keys, input, chunk, found_match, gsink.has_null);
+		break;
+	}
+	case JoinType::SEMI:
+		PhysicalJoin::ConstructSemiJoinResult(input, chunk, found_match);
+		break;
+	case JoinType::ANTI:
+		PhysicalJoin::ConstructAntiJoinResult(input, chunk, found_match);
+		break;
+	default:
+		throw NotImplementedException("Unimplemented join type for AsOf join");
+	}
+}
+
+OperatorResultType PhysicalAsOfJoin::ResolveComplexJoin(ExecutionContext &context, DataChunk &input, DataChunk &chunk,
+                                                        OperatorState &lstate_p) const {
+	auto &lstate = lstate_p.Cast<AsOfLocalState>();
+	auto &gsink = sink_state->Cast<AsOfGlobalSinkState>();
+
+	if (!lstate.fetch_next_left) {
+		lstate.fetch_next_left = true;
+		if (lstate.left_outer.Enabled()) {
+			// left join: before we move to the next chunk, see if we need to output any vectors that didn't
+			// have a match found
+			lstate.left_outer.ConstructLeftJoinResult(input, chunk);
+			lstate.left_outer.Reset();
+		}
+		return OperatorResultType::NEED_MORE_INPUT;
+	}
+
+	// perform the actual join
+	AsOfLocalState::Match matches[STANDARD_VECTOR_SIZE];
+	lstate.ResolveJoin(input, nullptr, matches);
+	lstate.group_payload.Reset();
+	lstate.rhs_payload.Reset();
+
+	auto &global_partition = gsink.global_partition;
+	hash_t scan_bin = global_partition.bin_groups.size();
+	PartitionGlobalHashGroup *hash_group = nullptr;
+	unique_ptr<PayloadScanner> scanner;
+	for (idx_t i = 0; i < lstate.lhs_match_count; ++i) {
+		const auto idx = lstate.lhs_matched[i];
+		const auto match_bin = matches[idx].first;
+		const auto match_pos = matches[idx].second;
+		if (match_bin != scan_bin) {
+			//	Grab the next group
+			const auto group_idx = global_partition.bin_groups[match_bin];
+			hash_group = global_partition.hash_groups[group_idx].get();
+			scan_bin = match_bin;
+			scanner = make_uniq<PayloadScanner>(*hash_group->global_sort, false);
+			lstate.group_payload.Reset();
+		}
+		// Skip to the range containing the match
+		while (match_pos >= scanner->Scanned()) {
+			lstate.group_payload.Reset();
+			scanner->Scan(lstate.group_payload);
```

Comment:
Profiling the benchmark almost all the time seems to be going here - in particular when I modify the benchmark to only join on the time column (`SELECT SUM(v) FROM probe ASOF JOIN build USING(t);`). It seems the random access is hurting performance.

Does it perhaps make sense to fully sort the left-side as well? Or perhaps to batch input from the left-side in larger buffers (e.g. 1M rows instead of 2K rows) and sorting that to reduce random access?

Reply:
Yeah I think there is still a fair bit of perf work to do, which I was planning to get to next. My big concern with this PR was to replace the old code, which was less capable, a lot slower and gave incorrect answers in some situations (probing with infinity would not match).

I was also thinking that something like what @lnkuiper did with hash joins, where a subset of the probe side that doesn't fit in memory is deferred for later matching might be good for scaling.

Reply:
You only have to defer a part of the probe side if the build side does not fit in memory. At least, that's how it works for the hash join. We Pin as much as we can on the build side, and probe whatever we can against that. Any tuples on the probe side that need to be compared against the part of the build side that is not pinned are deferred for later probing.

I think we can do something similar here, as you are partitioning your build side already. You can sink the tuples that _cannot_ match the currently pinned build side partitions into a `PartitionState`, and the tuples that _can_ match the currently pinned probe side can be probed immediately (or, as @Mytherin suggested, you can collect and sort batches).

If the build side is sufficiently small, batching/sorting is not necessary, as the build side will likely fit in L1/L2 cache. If the build side is big the random access will start hurting. I think this PR is good for now, though. The batching approach could be implemented like this: During `ExecuteInternal`, you collect, e.g., ~1M batches, then sort and probe. If the final call to `ExecuteInternal` has happened, and we still have some data saved up for the last batch, then the final sort/probe has to happen in `GetData` instead.